### ISSUE TITLE
[Bug] Unfixable scenario with reportUnusedCatchesOfUncheckedExceptions / [Breaking change request ?] Unchecked exception

### ISSUE THREAD

**Issue by VincentLanglet** [pepakriz/phpstan-exception-rules#116]
## Introduction

Considering the \LogicException was set as unchecked.

<details>
  <summary>Examples of error returned actually</summary>

When using this library, an error is reported for this code.
```
/**
 * @throws LogicException
 */
private function throwLogic(): void // error: Unused @throws LogicException annotation
{
    throw new LogicException();
}
```

An error is also reported without the option ignoreDescriptiveUncheckedExceptions for
```
/**
 * @throws LogicException Description
 */
private function throwLogic(): void // error: Unused @throws LogicException annotation
{
    throw new LogicException();
}
```

No error is reported for
```
private function throwLogic(): void
{
    throw new LogicException();
}
```

No error is reported without the option reportUnusedCatchesOfUncheckedExceptions for 
```
private function catchLogic(): void
{
    try {} catch (\LogicException $e) {}
}
```
</details>

## A non-fixable case

So now, we have a pretty weird situation: with `reportUnusedCatchesOfUncheckedExceptions: true`,
```
private function catchLogic(): void
{
    try {
        $this->throwLogic()
    } catch (\LogicException $e) {}
}

private function throwLogic(): void
{
    throw new LogicException();
}
```
Return an error (LogicException is never thrown in the corresponding try block). But
```
private function catchLogic(): void
{
    try {
        $this->throwLogic()
    } catch (\LogicException $e) {}
}

/**
 * @throws \LogicException
 */
private function throwLogic(): void
{
    throw new LogicException();
}
```
Return an error too (Unused @throws LogicException annotation).

<details>
  <summary> And there is worst </summary> 

```
private function catchLogic(): void
{
    try {
        $this->callThrowLogic()
    } catch (\LogicException $e) {} // Error useless catch
}

private function callThrowLogic(): void
{
    $this->throwLogic();
}

/**
 * @throws \LogicException
 */
private function throwLogic(): void // Error useless anotation
{
    throw new LogicException();
}
```
</details>

## Personal reflexion

First, since some linter are asking for an `@throws` tag for every exception, I feel like
```
/**
 * @throws LogicException
 */
private function throwLogic(): void
{
    throw new LogicException();
}
```
Should not return an error.
The `ignoreDescriptiveUncheckedExceptions` would not be needed anymore then.
You can annotate or not.

<details>
  <summary>But I think another option could be added to force the annotation for the function creating the exception</summary>

RequiredAnnotationForUncheckedExceptions.
```
/**
 * @throws LogicException
 */
private function throwLogic(): void
{
    throw new LogicException();
}

private function throwLogic2(): void // Error, the @throws is needed
{
    throw new LogicException();
}

private function callThrowLogic(): void // No error, the annotation is only needed for the function throwing the uncheckedException
{
    $this->throwLogic();
}
```

Some linters are trying that every Exception threw is documented ; but there is always false-positive/false-negative cause they only do static analysis. This is a classic false positive.
```
private function dontThrowLogic(): void // Error missing throw annotation
{
    try { throw new LogicException(); } catch (\LogicException $e) {}
}
```
</details >

But what is an unchecked exception then ?
For me, an unchecked exception should be an exception you don't have to catch. You can, but you don't have to.

So, both
```
private function callThrowLogic(): void
{
    $this->throwLogic();
}
```
and
```
private function callThrowLogic(): void
{
    try {
         $this->throwLogic();
    } catch (\LogicException $exception) {}
}
```
would be valid.

## How does it fix the issue

With `reportUnusedCatchesOfUncheckedExceptions: true`, the following code
```
private function catchLogic(): void
{
    try {
        $this->throwLogic()
    } catch (\LogicException $e) {}
}

private function throwLogic(): void
{
    throw new LogicException();
}
```
can now be fixed as 
```
private function catchLogic(): void
{
    try {
        $this->throwLogic()
    } catch (\LogicException $e) {}
}

/**
 * @throws \LogicException
 */
private function throwLogic(): void
{
    throw new LogicException();
}
```
without any error. This seems pretty normal for me: if you want to catch an exception (checked or unchecked) the function you called HAS TO document it.

**Comment by VincentLanglet** [comment]
@pepakriz I can try to work on this issue, but I need your point of view first.

**Comment by pepakriz** [comment]
Hi. Your idea is absolutely correct and I agree with your suggested solution. Maybe just one notice. In fact, there can be 3 types of exceptions: 1) checked 2) unchecked 3) errors. The third type should not be annotated at all.

I'd like to hear @marcospassos opinion because He has a similar idea and `ignoreDescriptiveUncheckedExceptions` is his work.

**Comment by VincentLanglet** [comment]
> Hi. Your idea is absolutely correct and I agree with your suggested solution.

Do you think it will be easy to implement ?

Needed: 
- 1° The following code should not return an error
```
/**
 * @throws LogicException
 */
private function throwLogic(): void
{
    throw new LogicException();
}
``` 
- 2° The option ignoreDescriptiveUncheckedExceptions should be removed



Facultative:
- 1° Adding the option RequiredAnnotationForUncheckedExceptions
- 2° Special case for Errors.

**Comment by VincentLanglet** [comment]
> I'd like to hear @marcospassos opinion because He has a similar idea and `ignoreDescriptiveUncheckedExceptions` is his work.

Hi, @pepakriz, it's been already 20 days. Should we start doing something about this unfixable scenario ?

**Comment by marcospassos** [comment]
HI @VincentLanglet, sorry for the delayed reply, busy days here.

I've always been in favor of allowing docs about unchecked exceptions. That's why I've implemented this feature in the first place. So, if @pepakriz agrees with it, I'm in.

**Comment by VincentLanglet** [comment]
> HI @VincentLanglet, sorry for the delayed reply, busy days here.

No problem ;)

> I've always been in favor of allowing docs about unchecked exceptions. That's why I've implemented this feature in the first place. So, if @pepakriz agrees with it, I'm in.

I'm happy to hear it !

Do you see how to implement theses changes @pepakriz ?

**Comment by VincentLanglet** [comment]
Hi @pepakriz, would you have time to fix this or at least to help me to fix this since I have no clue about how this code works ? :)

**Comment by pepakriz** [comment]
@VincentLanglet Can you prepare a PR with failing tests? Maybe someone'll take up the gauntlet.

**Comment by VincentLanglet** [comment]
@pepakriz I'll try.
But the test are already failing in master so I'm not sure it will help a lot.
Do you know why ?

**Comment by VincentLanglet** [comment]
@pepakriz Guess what ? Thanks to TDD I succeed to write the whole PR !
See https://github.com/pepakriz/phpstan-exception-rules/pull/119

Test are failing on the CI, but working locally when I installed with
```
composer update --prefer-lowest
```

I tried `prefer-lowest` since the tests were working before.
